### ISSUE TITLE
Full page flash when loading a prerendered page that uses lazy loading

### ISSUE THREAD

**Issue by danroth27** [dotnet/aspnetcore#25731]
Repro steps
- Install .NET 5 RC2 SDK: 5.0.100-rc.2.20459.6
- Clone https://github.com/danroth27/BlazorNet5Samples
- Run the app
- Click on the lazy loading tab
- Toggle the WebAssembly switch to run on WebAssembly

![7VXv0YNagF](https://user-images.githubusercontent.com/1874516/92622574-0c07ef80-f27a-11ea-82a4-7e207e385a52.gif)

**Comment by mihaimyh** [comment]
@danroth27 I was looking for an example of using server side Blazor and switching to Warm, is this what your sample is doing?

**Comment by danroth27** [comment]
>  I was looking for an example of using server side Blazor and switching to Wasm, is this what your sample is doing?

@mihaimyh Yeah, the app is setup so you can toggle back and forth between Blazor Server and Blazor WebAssembly at runtime.

**Comment by captainsafia** [comment]
This is reproable although I haven't been able to track down who is triggering the extra render.

Moving this to the backlog so that we can evaluate how many folks run into this in the wild.

**Comment by carlfranklin** [comment]
You can get rid of this by removing the <Navigating> section from App.razor. You can also add a little JavaScript to change the cursor. Add Id="xxx" to the NavLink objects in NavMenu.razor, then call JavaScript to set the cursor to "wait" before loading, and then again to "pointer" after loading. Works like a champ.

**Comment by GioviQ** [comment]
Also one should be able to decide in https://github.com/enkodellc/blazorboilerplate/blob/master/src/Shared/Modules/BlazorBoilerplate.Theme.Material/App.razor#L43
if "Navigating" should appear or not.

**Comment by javiercn** [comment]
@danroth27 there's an exception happening in your "gif".

It's not clear to me what's going on here, my theory is that you see the flash because the exception that it's happening. Could you see if this repros in 6.0? (and capture the exception).

The flashing is typically indicated by some async work going on (like fetching the dll) so we might be able to do something in 6.0 during prerendering to capture the list of assemblies that got lazy loaded and potentially eagerly load them when the app re-hydrates

**Comment by jsakamoto** [comment]
Recently I have run into this problem too.
I can reproduce this unexpected behavior on my minimal sample project (ASP.NET Core hosted Blazor WebAssembly app, with **lazy loading** and **server-side prerendering**) I have attached to this post.

- [📦BlazorPreRenderAndLazyLoadApp.zip](https://github.com/dotnet/aspnetcore/files/9102176/BlazorPreRenderAndLazyLoadApp.zip)

And also, I deployed that app on Azure Web Apps, so anybody can confirm this problem on the fly (The URL is below).

- https://demo-blazor-prerender-and-lazyload.azurewebsites.net/lazy-page

![](https://user-images.githubusercontent.com/95908/178736480-d0b58386-fcac-4b28-8fe1-77ecc460fba1.png)

<details>
<summary>and open here to show the comparison scenario:</summary>

![image](https://user-images.githubusercontent.com/95908/178737577-e136d55f-b4d1-4c09-87d9-c5abba24a480.png)

- [📕The full document of pictures above is here.](https://github.com/dotnet/aspnetcore/files/9102497/Report.pdf)
</details>

Maybe everybody can agree with me that implementing a `<Navigating>` render fragment in a `<Router>` component can not be a solution for this problem.
(That will only show something instead of empty contents. It doesn't resolve to flicker.)

I hope someone acknowledges this issue again and fixes it.

Or, as another idea, **can the rendering process during navigating be more customizable** by improving ASP.NET Core Blazor's code?

Actually, I've tried to reimplement a `<Router>`  component by myself to resolve this problem by suppressing rendering during lazy assemblies loading.
First, I copied the source code of the `<Router>` component of ASP.NET Core Blazor and tried to modify it. (It is an MIT license, right?)
However, the code of the `<Router>` component **strongly depends on internal members** of ASP.NET Core Blazor libraries, so I could not complete that attempt.

I'd really appreciate it if I could hack it myself to resolve this problem, even if the ASP.NET Core team would not fix it directly.

**Comment by fingers10** [comment]
@danroth27  .NET 7 release is few days ahead. Can we have any updates on this? Or this is planned for .NET 8?

**Comment by danroth27** [comment]
@fingers10 We haven't done any work on this for .NET 7. But we will evaluate the priority of this issue for .NET 8.

Please note that if an issue is important to you it's helpful to put a :+1: reaction on the **original issue post**. This helps us sort and prioritize issues that are most important to the community. Thank you for your input!

**Comment by SteveSandersonMS** [comment]
We'll track this in #45131, so closing as a dupe of that.
